### PR TITLE
Sf emails to s3  - save files in environment specific buckets

### PR DESCRIPTION
**Pull request by david-pepper**
## What does this change?
The changes in this PR enable the DEV and PROD versions of the Lambda to save emails to environment-specific buckets. 

Bucketname has been added as an element of S3 config.

The cloudformation file has been updated to create an environment variable for the bucketname which is suffixed by the environment e.g. emails-from-sf-dev, emails-from-sf-prod

### CODE REVIEW

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Config.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        userName = sfUserName,
+        clientId = sfClientId,
+        clientSecret = sfClientSecret,
+        password = sfPassword,
+        token = sfToken,
+        authUrl = sfAuthUrl,
+        apiVersion = sfApiVersion
+      ),
+      S3Config(
+        bucketName = s3BucketName + "-" + s3Stage.toLowerCase()
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Config.scala:45]:
```suggestion
        bucketName = s3BucketName
```

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Config.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    for {
+      sfUserName <- Option(System.getenv("username"))
+      sfClientId <- Option(System.getenv("clientId"))
+      sfClientSecret <- Option(System.getenv("clientSecret"))
+      sfPassword <- Option(System.getenv("password"))
+      sfToken <- Option(System.getenv("token"))
+      sfAuthUrl <- Option(System.getenv("authUrl"))
+      sfApiVersion <- Option(System.getenv("apiVersion"))
+      s3BucketName <- Option(System.getenv("bucketName"))
+      s3Stage <- Option(System.getenv("stage"))
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Config.scala:33]: Is this being used?

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala [hunk 3]
```diff
@@ -47,30 +47,31 @@ object Handler extends LazyLogging {
...
-      processNextPageOfEmails(sfAuthDetails, emailsDataFromSF.nextRecordsUrl.get)
+      processNextPageOfEmails(sfAuthDetails, emailsDataFromSF.nextRecordsUrl.get, bucketName)
     }
 
   }
 
-  def getEmailIdsSuccessfullySavedToS3(emailsDataFromSF: EmailsFromSfResponse.Response): Seq[String] = {
-
+  def getEmailIdsSuccessfullySavedToS3(emailsDataFromSF: EmailsFromSfResponse.Response, bucketName: String): Seq[String] = {
+    println("bucketName:" + bucketName)
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala:63]:
```suggestion
```

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala [hunk 4]
```diff
@@ -18,41 +18,39 @@ import scala.util.Try
 
 object S3Connector extends LazyLogging {
 
-  val bucketName = "emails-from-sf"
-
-  def generateJson(fileIsInS3: Boolean, caseEmail: EmailsFromSfResponse.Records): Either[CustomFailure, String] = {
+  def generateJson(fileIsInS3: Boolean, caseEmail: EmailsFromSfResponse.Records, bucketName: String): Either[CustomFailure, String] = {
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala:21]: If this function can't fail, it should probably just return a `String` and then lift that into an `Either` at the call site if necessary:
```suggestion
  def generateJson(fileIsInS3: Boolean, caseEmail: EmailsFromSfResponse.Records, bucketName: String): String = {
```

**Review by kelvin-chappell — COMMENTED** [review]
As discussed, could consider putting the SF auth into another type of `Config` class to reduce the number of arguments passed around but it's fine as it is.

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Config.scala (hunk 1 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Config.scala:45]: Yes you're right, as the bucket name will be appended with the stage by cloudformation when it gets created. This is done.

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Config.scala (hunk 2 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Config.scala:33]: No, I've removed it

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala (hunk 3 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala:63]: Removed

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala (hunk 4 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/S3Connector.scala]: This method was actually not being used anywhere, so I've removed it

**Review by kelvin-chappell — APPROVED** [review]
👍

**Pull request merged**